File: apps/server/providers/src/claude/convert.ts
/**
 * Converts AgentMessage[] / AgentTool[] to the Anthropic Messages wire
 * format (`POST /v1/messages`).
 *
 * Mapping:
 *   UserMessage       → role: "user", content blocks (text + images)
 *   AssistantMessage  → role: "assistant", content blocks (text + tool_use)
 *   ToolResultMessage → role: "user", content blocks (tool_result)
 */
import { zodToJsonSchema } from "zod-to-json-schema";
import type { AgentMessage, AgentTool, CacheRetention } from "@console/types";

export interface ClaudeMessage {
  role: "user" | "assistant";
  content: Array<Record<string, unknown>>;
}

export interface ClaudeTool {
  name: string;
  description?: string;
  input_schema: Record<string, unknown>;
  cache_control?: { type: "ephemeral" };
}

/** Placeholder for empty tool results — the Messages API rejects empty content. */
export const EMPTY_TOOL_RESULT_TEXT = "Tool failed with no output.";

function toolResultText(content: unknown): string {
  const text = typeof content === "string" ? content : JSON.stringify(content);
  return text.trim() === "" ? EMPTY_TOOL_RESULT_TEXT : text;
}

function parseToolInput(input: unknown): Record<string, unknown> {
  if (input && typeof input === "object" && !Array.isArray(input)) {
    return input as Record<string, unknown>;
  }
  if (typeof input === "string" && input.trim() !== "") {
    try {
      const parsed: unknown = JSON.parse(input);
      if (parsed && typeof parsed === "object" && !Array.isArray(parsed)) {
        return parsed as Record<string, unknown>;
      }
    } catch {
      // Fall through to empty input below.
    }
  }
  return {};
}

function normalizeImageMime(mimeType: string): "image/jpeg" | "image/png" | "image/gif" | "image/webp" | undefined {
  const normalized = mimeType.trim().toLowerCase();
  if (normalized === "image/jpg") return "image/jpeg";
  if (
    normalized === "image/jpeg" ||
    normalized === "image/png" ||
    normalized === "image/gif" ||
    normalized === "image/webp"
  ) {
    return normalized;
  }
  return undefined;
}

export function convertClaudeMessages(messages: AgentMessage[], cacheRetention?: CacheRetention): ClaudeMessage[] {
  const turns: ClaudeMessage[] = [];

  for (const msg of messages) {
    if (msg.role === "user") {
      const text = msg.content ?? "";
      const attachments = msg.attachments ?? [];
      if (text.trim() === "" && attachments.length === 0) continue;
      const content: Array<Record<string, unknown>> = [];
      if (text.trim() !== "") {
        content.push({ type: "text", text });
      } else {
        content.push({ type: "text", text: "(see attached images)" });
      }
      for (const att of attachments) {
        const mediaType = normalizeImageMime(att.mimeType);
        if (!mediaType) {
          content.push({ type: "text", text: `[unsupported image: ${att.mimeType}]` });
          continue;
        }
        content.push({
          type: "image",
          source: { type: "base64", media_type: mediaType, data: att.data },
        });
      }
      turns.push({ role: "user", content });
      continue;
    }

    if (msg.role === "assistant") {
      const content: Array<Record<string, unknown>> = [];
      for (const part of msg.content) {
        if (part.type === "text" && part.text) {
          content.push({ type: "text", text: part.text });
        } else if (part.type === "thinking" && part.text) {
          // Prior thinking is replayed unsigned, which the signing endpoint
          // rejects — demote it to text (mirrors oh-my-pi's demotion path)
          // instead of dropping the reasoning context entirely.
          content.push({ type: "text", text: part.text });
        } else if (part.type === "toolCall") {
          content.push({
            type: "tool_use",
            id: part.call.id,
            name: part.call.name,
            input: parseToolInput(part.call.arguments),
          });
        }
      }
      if (content.length > 0) {
        turns.push({ role: "assistant", content });
      }
      continue;
    }

    if (msg.role === "toolResult") {
      const content: Array<Record<string, unknown>> = msg.results.map((r) => ({
        type: "tool_result",
        tool_use_id: r.toolCallId,
        content: toolResultText(r.content),
        ...(r.isError ? { is_error: true } : {}),
      }));
      if (content.length > 0) {
        // The Anthropic wire keys results by tool_use_id only.
        turns.push({ role: "user", content });
      }
    }
  }

  // Merge adjacent same-role turns (Anthropic requires strict alternation).
  const merged: ClaudeMessage[] = [];
  for (const turn of turns) {
    const last = merged[merged.length - 1];
    if (last && last.role === turn.role) {
      last.content.push(...turn.content);
    } else {
      merged.push({ role: turn.role, content: [...turn.content] });
    }
  }

  // Drop leading assistant turns; never send an empty conversation.
  while (merged.length > 0 && merged[0]!.role !== "user") {
    merged.shift();
  }
  if (merged.length === 0) {
    merged.push({ role: "user", content: [{ type: "text", text: "(continue)" }] });
  }

  // Trailing prompt-cache breakpoint over the conversation prefix, so the
  // stable history caches across turns. Skipped when the caller opts out.
  if (cacheRetention !== "none") {
    const lastMessage = merged[merged.length - 1]!;
    const lastBlock = lastMessage.content[lastMessage.content.length - 1];
    if (lastBlock) {
      lastBlock.cache_control = { type: "ephemeral" };
    }
  }

  return merged;
}

/**
 * Anthropic consumes draft-2020-12 JSON Schema. zod-to-json-schema's OpenAPI
 * output still uses draft-07's boolean exclusiveMinimum/exclusiveMaximum
 * keywords, which strict providers reject — same upgrade Codex applies.
 */
export function normalizeClaudeSchema(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(normalizeClaudeSchema);
  if (!value || typeof value !== "object") return value;

  const schema = value as Record<string, unknown>;
  const normalized: Record<string, unknown> = {};
  for (const [key, child] of Object.entries(schema)) {
    normalized[key] = normalizeClaudeSchema(child);
  }

  if (normalized.exclusiveMinimum === true) {
    if (typeof normalized.minimum === "number") normalized.exclusiveMinimum = normalized.minimum;
    else delete normalized.exclusiveMinimum;
  } else if (normalized.exclusiveMinimum === false) {
    delete normalized.exclusiveMinimum;
  }
  if (normalized.exclusiveMaximum === true) {
    if (typeof normalized.maximum === "number") normalized.exclusiveMaximum = normalized.maximum;
    else delete normalized.exclusiveMaximum;
  }

  return normalized;
}

export function convertClaudeTools(tools: AgentTool[], cacheRetention?: CacheRetention): ClaudeTool[] {
  return tools.map((tool, index) => {
    const rawSchema = zodToJsonSchema(tool.inputSchema, {
      target: "openApi3",
      $refStrategy: "none",
    }) as Record<string, unknown>;
    const input_schema =
      rawSchema?.type === "object"
        ? (normalizeClaudeSchema(rawSchema) as Record<string, unknown>)
        : { type: "object" };
    return {
      name: tool.name,
      ...(tool.description ? { description: tool.description } : {}),
      input_schema,
      // Cache the (static) tool definitions with a breakpoint on the last
      // tool. Skipped when the caller opts out of caching.
      ...(cacheRetention !== "none" && index === tools.length - 1
        ? { cache_control: { type: "ephemeral" } }
        : {}),
    };
  });
}

File: apps/server-go/internal/providers/claude/convert.go
// Claude wire conversion: loop messages/tools to the Anthropic Messages
// format. Port of apps/server/providers/src/claude/convert.ts.
package claude

import (
	"encoding/json"
	"strings"

	"github.com/Adelodunpeter25/console/apps/server-go/internal/agent/loop"
	"github.com/Adelodunpeter25/console/apps/server-go/internal/agent/tools"
)

// ClaudeMessage is one Anthropic conversation turn.
type ClaudeMessage struct {
	Role    string           `json:"role"`
	Content []map[string]any `json:"content"`
}

// ClaudeTool is one Anthropic tool definition.
type ClaudeTool struct {
	Name         string         `json:"name"`
	Description  string         `json:"description,omitempty"`
	InputSchema  map[string]any `json:"input_schema"`
	CacheControl map[string]any `json:"cache_control,omitempty"`
}

// EmptyToolResultText is the placeholder for empty tool results — the
// Messages API rejects empty content.
const EmptyToolResultText = "Tool failed with no output."

func toolResultText(content any) string {
	var text string
	if s, ok := content.(string); ok {
		text = s
	} else {
		raw, _ := json.Marshal(content)
		text = string(raw)
	}
	if strings.TrimSpace(text) == "" {
		return EmptyToolResultText
	}
	return text
}

func parseToolInput(input any) map[string]any {
	if m, ok := input.(map[string]any); ok {
		return m
	}
	if s, ok := input.(string); ok && strings.TrimSpace(s) != "" {
		var m map[string]any
		if err := json.Unmarshal([]byte(s), &m); err == nil {
			return m
		}
	}
	if raw, ok := input.(json.RawMessage); ok && len(raw) > 0 {
		var m map[string]any
		if err := json.Unmarshal(raw, &m); err == nil {
			return m
		}
	}
	return map[string]any{}
}

func normalizeImageMime(mimeType string) string {
	switch strings.TrimSpace(strings.ToLower(mimeType)) {
	case "image/jpg":
		return "image/jpeg"
	case "image/jpeg", "image/png", "image/gif", "image/webp":
		return strings.TrimSpace(strings.ToLower(mimeType))
	default:
		return ""
	}
}

// partText extracts display text from an assistant content part in any of
// the Go shapes (typed structs or decoded maps).
func partText(part any) (text string, isThinking bool) {
	switch p := part.(type) {
	case loop.TextPart:
		return p.Text, false
	case loop.ThinkingPart:
		return p.Text, true
	case map[string]any:
		if t, _ := p["text"].(string); t != "" {
			if p["type"] == "thinking" {
				return t, true
			}
			return t, false
		}
	}
	return "", false
}

// partToolCall extracts a tool call from an assistant content part.
func partToolCall(part any) (id, name string, args any, ok bool) {
	switch p := part.(type) {
	case loop.ToolCallPart:
		return p.Call.ID, p.Call.Name, p.Call.Arguments, true
	case map[string]any:
		if p["type"] != "toolCall" {
			return "", "", nil, false
		}
		call, _ := p["call"].(map[string]any)
		if call == nil {
			return "", "", nil, false
		}
		id, _ := call["id"].(string)
		name, _ := call["name"].(string)
		return id, name, call["arguments"], id != "" && name != ""
	}
	return "", "", nil, false
}

// ConvertMessages maps loop history to Anthropic turns: strict role
// alternation, orphan-result sanitizing, and a trailing cache breakpoint.
func ConvertMessages(messages []any, retention loop.CacheRetention) []ClaudeMessage {
	turns := make([]ClaudeMessage, 0, len(messages))
	for _, m := range messages {
		switch msg := m.(type) {
		case loop.UserMessage:
			text := msg.Content
			attachments := msg.Attachments
			if strings.TrimSpace(text) == "" && len(attachments) == 0 {
				continue
			}
			content := make([]map[string]any, 0, len(attachments)+1)
			if strings.TrimSpace(text) != "" {
				content = append(content, map[string]any{"type": "text", "text": text})
			} else {
				content = append(content, map[string]any{"type": "text", "text": "(see attached images)"})
			}
			for _, att := range attachments {
				mediaType := normalizeImageMime(att.MimeType)
				if mediaType == "" {
					content = append(content, map[string]any{
						"type": "text", "text": "[unsupported image: " + att.MimeType + "]",
					})
					continue
				}
				content = append(content, map[string]any{
					"type": "image",
					"source": map[string]any{
						"type": "base64", "media_type": mediaType, "data": att.Data,
					},
				})
			}
			turns = append(turns, ClaudeMessage{Role: "user", Content: content})
		case *loop.UserMessage:
			if msg != nil {
				turns = append(turns, ConvertMessages([]any{*msg}, retention)...)
			}
		case loop.AssistantMessage:
			var texts, toolUses []map[string]any
			for _, part := range msg.Content {
				if text, thinking := partText(part); text != "" {
					// Prior thinking replays unsigned, which the signing
					// endpoint rejects — demote to text instead of dropping.
					_ = thinking
					texts = append(texts, map[string]any{"type": "text", "text": text})
				} else if id, name, args, ok := partToolCall(part); ok {
					toolUses = append(toolUses, map[string]any{
						"type": "tool_use", "id": id, "name": name, "input": parseToolInput(args),
					})
				}
			}
			// All text must precede tool_use blocks in an Anthropic turn.
			content := append(texts, toolUses...)
			if len(content) > 0 {
				turns = append(turns, ClaudeMessage{Role: "assistant", Content: content})
			}
		case loop.ToolResultMessage:
			content := make([]map[string]any, 0, len(msg.Results))
			for _, r := range msg.Results {
				block := map[string]any{
					"type": "tool_result", "tool_use_id": r.ToolCallID,
					"content": toolResultText(r.Content),
				}
				if r.IsError {
					block["is_error"] = true
				}
				content = append(content, block)
			}
			if len(content) > 0 {
				turns = append(turns, ClaudeMessage{Role: "user", Content: content})
			}
		}
	}

	// Merge adjacent same-role turns (Anthropic requires strict alternation).
	merged := make([]ClaudeMessage, 0, len(turns))
	for _, turn := range turns {
		if n := len(merged); n > 0 && merged[n-1].Role == turn.Role {
			merged[n-1].Content = append(merged[n-1].Content, turn.Content...)
		} else {
			merged = append(merged, ClaudeMessage{Role: turn.Role, Content: append([]map[string]any{}, turn.Content...)})
		}
	}

	// Conversations must start and end with a user turn.
	if len(merged) == 0 {
		merged = append(merged, ClaudeMessage{Role: "user", Content: []map[string]any{{"type": "text", "text": "(continue)"}}})
	} else {
		if merged[0].Role != "user" {
			merged = append([]ClaudeMessage{{Role: "user", Content: []map[string]any{{"type": "text", "text": "(session started)"}}}}, merged...)
		}
		if merged[len(merged)-1].Role != "user" {
			merged = append(merged, ClaudeMessage{Role: "user", Content: []map[string]any{{"type": "text", "text": "(continue)"}}})
		}
	}

	// Sanitize tool results: every tool_result needs a matching tool_use in
	// the immediately preceding assistant turn; orphans become text.
	for i := range merged {
		if merged[i].Role != "user" {
			continue
		}
		valid := map[string]bool{}
		if i > 0 && merged[i-1].Role == "assistant" {
			for _, block := range merged[i-1].Content {
				if block["type"] == "tool_use" {
					if id, _ := block["id"].(string); id != "" {
						valid[id] = true
					}
				}
			}
		}
		for j, block := range merged[i].Content {
			if block["type"] != "tool_result" {
				continue
			}
			id, _ := block["tool_use_id"].(string)
			if valid[id] {
				continue
			}
			text := ""
			if s, ok := block["content"].(string); ok {
				text = s
			} else {
				raw, _ := json.Marshal(block["content"])
				text = string(raw)
			}
			merged[i].Content[j] = map[string]any{"type": "text", "text": "[Tool result: " + text + "]"}
		}
	}

	// Trailing cache breakpoint over the conversation prefix. Skipped when
	// the caller opts out.
	if retention != loop.CacheNone && len(merged) > 0 {
		last := &merged[len(merged)-1]
		if n := len(last.Content); n > 0 {
			last.Content[n-1]["cache_control"] = map[string]any{"type": "ephemeral"}
		}
	}
	return merged
}

// NormalizeSchema upgrades draft-07 boolean exclusiveMinimum/Maximum to
// draft-2020-12 numeric form (same upgrade Codex applies).
func NormalizeSchema(value any) any {
	switch v := value.(type) {
	case []any:
		out := make([]any, len(v))
		for i, item := range v {
			out[i] = NormalizeSchema(item)
		}
		return out
	case map[string]any:
		normalized := make(map[string]any, len(v))
		for key, child := range v {
			normalized[key] = NormalizeSchema(child)
		}
		if em, ok := normalized["exclusiveMinimum"]; ok {
			if em == true {
				if minimum, ok := normalized["minimum"].(float64); ok {
					normalized["exclusiveMinimum"] = minimum
				} else {
					delete(normalized, "exclusiveMinimum")
				}
			} else if em == false {
				delete(normalized, "exclusiveMinimum")
			}
		}
		if em, ok := normalized["exclusiveMaximum"]; ok {
			if em == true {
				if maximum, ok := normalized["maximum"].(float64); ok {
					normalized["exclusiveMaximum"] = maximum
				} else {
					delete(normalized, "exclusiveMaximum")
				}
			} else if em == false {
				delete(normalized, "exclusiveMaximum")
			}
		}
		return normalized
	default:
		return value
	}
}

// ConvertTools maps tool definitions to Anthropic tools, with a cache
// breakpoint on the last (static) definition unless opted out.
func ConvertTools(defs []tools.Definition, retention loop.CacheRetention) []ClaudeTool {
	out := make([]ClaudeTool, 0, len(defs))
	for i, t := range defs {
		schema := map[string]any{"type": "object"}
		if t.InputSchema != nil {
			if normalized, ok := NormalizeSchema(t.InputSchema).(map[string]any); ok {
				schema = normalized
			}
		}
		tool := ClaudeTool{Name: t.Name, InputSchema: schema}
		if t.Description != "" {
			tool.Description = t.Description
		}
		if retention != loop.CacheNone && i == len(defs)-1 {
			tool.CacheControl = map[string]any{"type": "ephemeral"}
		}
		out = append(out, tool)
	}
	return out
}
